geth/peers/peerpool.go

```diff
@@ -0,0 +1,278 @@
+	quit          chan struct{}
+
+	wg sync.WaitGroup
+}
+
+// init creates data structures used by peer pool
+// thread safety must be guaranteed by caller
+func (p *PeerPool) init() {
+	p.subscriptions = make([]event.Subscription, 0, len(p.config))
+	p.syncPeriods = make([]chan time.Duration, 0, len(p.config))
+	p.peers = make(map[discv5.Topic]map[discv5.NodeID]*peerInfo)
+	for topic := range p.config {
+		p.peers[topic] = make(map[discv5.NodeID]*peerInfo)
+	}
+}
+
+// Start creates discovery search query for each topic and consumes peers found in that topic
+// in separate loop.
+func (p *PeerPool) Start(server *p2p.Server) error {
+	if server.DiscV5 == nil {
+		return ErrDiscv5NotRunning
+	}
+	p.mu.Lock()
+	defer p.mu.Unlock()
+	p.quit = make(chan struct{})
+	// sync periods are stored because we need to close them once pool is stopped
+	p.init()
+
+	// 2 goroutines per each topic
+	p.wg.Add(len(p.config) * 2)
+	for topic, limits := range p.config {
+		topic := topic
+		period := make(chan time.Duration, 2)
+		p.syncPeriods = append(p.syncPeriods, period)
+		found := make(chan *discv5.Node, 10)
+		if p.cache != nil {
+			for _, peer := range p.cache.GetPeersRange(topic, 5) {
+				log.Debug("adding a peer from cache", "peer", peer)
+				found <- peer
+			}
+		}
+		lookup := make(chan bool, 100)
+		events := make(chan *p2p.PeerEvent, 20)
+		subscription := server.SubscribeEvents(events)
+		p.subscriptions = append(p.subscriptions, subscription)
+		log.Debug("running peering for", "topic", topic, "limits", limits)
+		go func() {
+			server.DiscV5.SearchTopic(topic, period, found, lookup)
+			p.wg.Done()
+		}()
+
+		go func() {
+			p.handlePeersFromTopic(server, topic, period, found, lookup, events)
+			p.wg.Done()
+		}()
+	}
+	return nil
+}
+
+func (p *PeerPool) handlePeersFromTopic(server *p2p.Server, topic discv5.Topic, period chan<- time.Duration, found <-chan *discv5.Node, lookup <-chan bool, events <-chan *p2p.PeerEvent) {
+	limits := p.config[topic]
+	fast := true
+	period <- p.fastSync
+	connected := 0
+	selfID := discv5.NodeID(server.Self().ID)
+	for {
+		select {
+		case <-p.quit:
+			return
+		case node := <-found:
+			log.Debug("found node with", "ID", node.ID, "topic", topic)
+			if node.ID == selfID {
+				continue
+			}
+			if p.processFoundNode(server, connected, topic, node) {
+				connected++
+			}
+			// switch period only once
+			if fast && connected >= limits[0] {
+				log.Debug("switch to slow sync", "topic", topic, "sync", p.slowSync)
+				period <- p.slowSync
+				fast = false
+			}
+		case <-lookup:
+			// just drain this channel for now, it can be used to intelligently
+			// limit number of kademlia lookups
+		case event := <-events:
+			if event.Type == p2p.PeerEventTypeDrop {
+				log.Debug("node dropped", "ID", event.Peer, "topic", topic)
+				if !p.processDroppedNode(server, topic, event.Peer) {
+					connected--
+				}
+				// switch period only once
+				if !fast && connected < limits[0] {
```

Comment:
How is it done only once? It looks like it can switch between fast and slow many times.

Reply:
i meant that new timer will be sent only once per switch, e.g. it is ok to send slow, fast, slow. but it is unnecessary to send slow,slow,slow